### ISSUE TITLE
Compute instances with custom machine types fail to update if the machine type maps to a standard size

### ISSUE THREAD

**Issue by joshheinrichs-shopify** [GoogleCloudPlatform/k8s-config-connector#829]
### Checklist

- [X] I did not find a related open issue.
- [X] I did not find a solution in the troubleshooting guide: (https://cloud.google.com/config-connector/docs/troubleshooting)
- [X] If this issue is time-sensitive, I have submitted a corresponding issue with [GCP support](https://cloud.google.com/support-hub).

### Bug Description

If you create a compute instance with a custom machine type that happens to line up with a standard machine type, e.g. `n2-custom-2-16384`, which lines up with `n2-highmem-2`, the instance will be created with a machine type of `n2-highmem-2`, and configconnector will run into this error because it thinks the instance's machine type needs updating:
```json
"conditions": [
    {
        "lastTransitionTime": "2023-06-21T17:30:44Z",
        "message": "Update call failed: error applying desired state: summary: Changing the machine_type, min_cpu_platform, service_account, enable_display, shielded_instance_config, scheduling.node_affinities or network_interface.[#d].(network/subnetwork/subnetwork_project) or advanced_machine_features on a started instance requires stopping it. To acknowledge this, please set allow_stopping_for_update = true in your config. You can also stop it by setting desired_status = \"TERMINATED\", but the instance will not be restarted after the update.",
        "reason": "UpdateFailed",
        "status": "False",
        "type": "Ready"
    }
],
```

### Additional Diagnostic Information

n/a

### Kubernetes Cluster Version

v1.24.12-gke.1000

### Config Connector Version

1.89.0

### Config Connector Mode

cluster mode

### Steps to reproduce the issue

* create a computeinstance with a custom machine type, e.g. `n2-custom-2-16384`

**Comment by diviner524** [comment]
@joshheinrichs-shopify This seems to be a problematic behavior in the underlying GCP API. Config Connector in this case is treating the difference between `n2-custom-2-16384` and `n2-highmem-2` as a drift thus trying to correct it.

I assume it works fine if we specify `n2-highmem-2` in the initial create request?

**Comment by joshheinrichs-shopify** [comment]
Yeah, if you specify `n2-highmem-2` things work fine. I worked around the issue by grabbing all the standard machine types via https://cloud.google.com/compute/docs/reference/rest/v1/machineTypes/list and mapping custom machine types to standard machine types when possible, but it'd be nice if that wasn't needed.

**Comment by diviner524** [comment]
@joshheinrichs-shopify Thanks for confirming! Would you be able to collect and share some sample HTTP requests/responses for the GCP API [1], which demonstrates the returned `machineType` is different than the value specified in the original request? 

You can consider filing a GCP support case to suggest improvement in the GCP API.

We can also help to provide some feedback to the Compute Engine team internally, and see if they can fix this.

[1] https://cloud.google.com/compute/docs/reference/rest/v1/instances

**Comment by joshheinrichs-shopify** [comment]
Sure! I redacted a bunch of stuff but this should get the point across:

I sent a POST to https://www.googleapis.com/compute/v1/projects/{project}/zones/us-central1-a/instances with
```
{
  "machineType": "projects/{project}/zones/us-central1-a/machineTypes/n2-custom-2-8192",
  "name": "test",
  ...
}
```
The response is an op which I'm guessing you don't care about. The eventual instance has a different machine type:
```
{
  "machineType": "projects/{project}/zones/us-central1-a/machineTypes/n2-standard-2",
  "name": "test",
  "zone": "projects/{project}/zones/us-central1-a",
  ...
}
```
There was no mention of `n2-custom-2-8192` anywhere.

Looking at the actual machine types via machineTypes.get https://cloud.google.com/compute/docs/reference/rest/v1/machineTypes/get, `n2-custom-2-8192` and `n2-standard-2` appear to be distinct types with different IDs:
```
{
  "kind": "compute#machineType",
  "id": "867583401984",
  "name": "n2-custom-2-8192",
  "description": "Custom created machine type.",
  "guestCpus": 2,
  "memoryMb": 8192,
  "maximumPersistentDisks": 128,
  "maximumPersistentDisksSizeGb": "263168",
  "zone": "us-central1-a",
  "selfLink": "https://www.googleapis.com/compute/v1/projects/{project}/zones/us-central1-a/machineTypes/n2-custom-2-8192",
  "isSharedCpu": false
}
```
```
{
  "kind": "compute#machineType",
  "id": "901002",
  "creationTimestamp": "1969-12-31T16:00:00.000-08:00",
  "name": "n2-standard-2",
  "description": "2 vCPUs 8 GB RAM",
  "guestCpus": 2,
  "memoryMb": 8192,
  "imageSpaceGb": 0,
  "maximumPersistentDisks": 128,
  "maximumPersistentDisksSizeGb": "263168",
  "zone": "us-central1-a",
  "selfLink": "https://www.googleapis.com/compute/v1/projects/{project}/zones/us-central1-a/machineTypes/n2-standard-2",
  "isSharedCpu": false
}
```

Setting the machine type to `n2-custom-2-8192` via https://cloud.google.com/compute/docs/reference/rest/v1/instances/setMachineType also results in `n2-standard-2`.

One potential complication is that custom machine types typically cost more than standard machine types, so although this behaviour is confusing it may actually be beneficial for some customers.

**Comment by newtondev** [comment]
The reference issue on https://github.com/hashicorp/terraform-provider-google/issues/12944 is marked as closed and I see a few merge requests that were merged.

@diviner524, is this something I could possibly pick up and work on?
